File: programmers/150366.cpp
#include <string>
#include <vector>
#include <sstream>

using namespace std;

#define N 51

int parents[N * N];
string values[N * N];

int getParents(int x) {
    if (parents[x] == x) return x;
    else return parents[x] = getParents(parents[x]);
}

void update1(int r, int c, string value) {
    int k = r * 50 + c;
    
    values[getParents(k)] = value;
}

void update2(string value1, string value2) {
    for (int i = 1; i < N * N; i++) {
        if (values[i] == value1) values[i] = value2;
    }
}

void merge(int r1, int c1, int r2, int c2) {
    int a = getParents(r1 * 50 + c1);
    int b = getParents(r2 * 50 + c2);
    
    if (a == b) return;
    if (values[a] == "" && values[b] != "") parents[a] = b;
    else parents[b] = a;
}

void unmerge(int r, int c) {
    int k = r * 50 + c;
    int chk = getParents(k);
    string value = values[chk];
    vector<int> v;
    
    for (int i = 1; i < N; i++) {
        for (int j = 1; j < N; j++) {
            int k = i * 50 + j;
            if (getParents(k) == chk) v.push_back(k);
        }
    }
    for (int cur : v) {
        parents[cur] = cur;
        values[cur] = "";
    }
    values[k] = value;
}

vector<string> solution(vector<string> commands) {
    vector<string> answer;
    
    for (int i = 1; i < N * N; i++) parents[i] = i;
    
    for (string command: commands) {
        istringstream iss(command);
        vector<string> tokens;
        string token;
        
        while (iss >> token) tokens.push_back(token);
        
        if (tokens[0] == "UPDATE") {
            if (tokens.size() == 4) update1(stoi(tokens[1]), stoi(tokens[2]), tokens[3]);
            else update2(tokens[1], tokens[2]);
        }
        else if (tokens[0] == "MERGE") {
            merge(stoi(tokens[1]), stoi(tokens[2]), stoi(tokens[3]), stoi(tokens[4]));
        }
        else if (tokens[0] == "UNMERGE") {
            unmerge(stoi(tokens[1]), stoi(tokens[2]));
        }
        else {
            int r = stoi(tokens[1]);
            int c = stoi(tokens[2]);
            int k = r * 50 + c;
            
            if (values[getParents(k)] == "") answer.push_back("EMPTY");
            else answer.push_back(values[getParents(k)]);
        }
/*
        for (int i = 1; i < 5; i++) {
            for (int j = 1; j < 5 ; j++) {
                printf("%s ", values[getParents(i * 50 + j)].c_str());
            }
            printf("\n");
        }
        printf("\n");*/
    }
    
    return answer;
}

File: programmers/150366.java
import java.util.*;

class Solution {
    private final static int N = 51;
    
    private String[][] table = new String[N][N];
    private int[][] id = new int[N][N];
    
    public String[] solution(String[] commands) {
        List<String> result = new ArrayList<>();
        
        for (int i = 1; i < N; i++) {
            for (int j = 1; j < N; j++) {
                id[i][j] = i * N + j;
            }
        }
        
        for (String cmd : commands) {
            String[] tokens = cmd.split(" ");
            
            switch(tokens[0]) {
                    case "UPDATE" -> update(tokens);
                    case "MERGE" -> merge(tokens);
                    case "UNMERGE" -> unmerge(tokens);
                    case "PRINT" -> result.add(print(tokens));
            }
        }
        
        return result.toArray(String[]::new);
    }
    
    private void update(String[] tokens) {
        if (tokens.length == 4) {
            int r = Integer.parseInt(tokens[1]);
            int c = Integer.parseInt(tokens[2]);
            int targetId = id[r][c];
            String value = tokens[3];

            for (int i = 1; i < N; i++) {
                for (int j = 1; j < N; j++) {
                    if (id[i][j] == targetId) {
                        table[i][j] = value;
                    }
                }
            }
        } else if (tokens.length == 3) {
            String value1 = tokens[1];
            String value2 = tokens[2];
            
            for (int i = 1; i < N; i++) {
                for (int j = 1; j < N; j++) {
                    if (table[i][j] != null && table[i][j].equals(value1)) {
                        table[i][j] = value2;
                    }
                }
            }
        }
    }
    
    private void merge(String[] tokens) {
        if (tokens.length != 5) return;
        
        int r1 = Integer.parseInt(tokens[1]);
        int c1 = Integer.parseInt(tokens[2]);
        int r2 = Integer.parseInt(tokens[3]);
        int c2 = Integer.parseInt(tokens[4]);
        int targetId1 = id[r1][c1];
        int targetId2 = id[r2][c2];
        
        if (targetId1 == targetId2) return;
        
        String value = (table[r1][c1] == null) ? table[r2][c2] : table[r1][c1];
        
        for (int i = 1; i < N; i++) {
            for (int j = 1; j < N; j++) {
                if (id[i][j] == targetId2) {
                    id[i][j] = targetId1;
                }
                if (id[i][j] == targetId1) {
                    table[i][j] = value;
                }
            }
        }
    }
    
    private void unmerge(String[] tokens) {
        if (tokens.length != 3) return;
        
        int r = Integer.parseInt(tokens[1]);
        int c = Integer.parseInt(tokens[2]);
        int targetId = id[r][c];
        String value = table[r][c];
        
        for (int i = 1; i < N; i++) {
            for (int j = 1; j < N; j++) {
                if (id[i][j] == targetId) {
                    id[i][j] = i * N + j;
                    table[i][j] = null;
                }
            }
        }
        table[r][c] = value;
    }
    
    private String print(String[] tokens) {
        if (tokens.length != 3) return null;
        
        int r = Integer.parseInt(tokens[1]);
        int c = Integer.parseInt(tokens[2]);
        
        return (table[r][c] == null) ? "EMPTY" : table[r][c];
    }
}